core/server/src/main/java/alluxio/master/block/BlockMaster.java

```diff
@@ -573,33 +603,32 @@ public void workerRegister(long workerId, List<String> storageTiers,
    */
   public Command workerHeartbeat(long workerId, Map<String, Long> usedBytesOnTiers,
       List<Long> removedBlockIds, Map<String, List<Long>> addedBlocksOnTiers) {
-    synchronized (mBlocks) {
-      synchronized (mWorkers) {
-        if (!mWorkers.contains(mIdIndex, workerId)) {
-          LOG.warn("Could not find worker id: {} for heartbeat.", workerId);
-          return new Command(CommandType.Register, new ArrayList<Long>());
-        }
+    MasterWorkerInfo workerInfo = mWorkers.getFirstByField(mIdIndex, workerId);
+    if (workerInfo == null) {
+      LOG.warn("Could not find worker id: {} for heartbeat.", workerId);
+      return new Command(CommandType.Register, new ArrayList<Long>());
+    }
 
-        MasterWorkerInfo workerInfo = mWorkers.getFirstByField(mIdIndex, workerId);
-        processWorkerRemovedBlocks(workerInfo, removedBlockIds);
-        processWorkerAddedBlocks(workerInfo, addedBlocksOnTiers);
+    synchronized (workerInfo) {
+      processWorkerRemovedBlocks(workerInfo, removedBlockIds);
```

Comment:
Technically, we do need to check. However, the only way this can be removed, is if it was detected as a lost worker. That means, it was transferred to the lost worker map. This scenario can happen if the worker heartbeats right as it is being considered as lost.

I originally thought this was ok, because it would heartbeat, but the next time it tries to heartbeat, it would be considered lost, and would be asked to re-register.

Essentially, I intentionally opted for not totally ordering the heartbeat and lost worker detection, since the worker would be asked to re-register either in the first heartbeat or the next.